packages/wdio-browserstack-service/src/util.ts

```diff
@@ -67,3 +82,338 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+            const passedExamples = child.scenario.examples.flatMap((val) => (val.tableBody)).find((item) => item.id == examplesId)?.cells.map((val) => (val.value))
+            if (passedExamples) examples = passedExamples
+        }
+    })
+
+    if (examples.length) return examples
+    return
+}
+
+export function removeAnsiColors(message: string): string {
+    // https://stackoverflow.com/a/29497680
+    // eslint-disable-next-line no-control-regex
+    return message.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '')
+}
+
+export function getLogTag(eventType: string): string {
+    if (eventType == 'TestRunStarted' || eventType == 'TestRunFinished') {
+        return 'Test_Upload'
+    } else if (eventType == 'HookRunStarted' || eventType == 'HookRunFinished') {
+        return 'Hook_Upload'
+    } else if (eventType == 'ScreenshotCreated') {
+        return 'Screenshot_Upload'
+    } else if (eventType == 'LogCreated') {
+        return 'Log_Upload'
+    }
+    return 'undefined'
+}
+
+function keepAliveAgent () {
+    return {
+        http: new http.Agent({ keepAlive: true }),
+        https: new https.Agent({ keepAlive: true }),
+    }
+}
+
+export async function uploadEventData (eventData: UploadType) {
+    const logTag: string = getLogTag(eventData.event_type)
+
+    if (eventData.event_type == 'ScreenshotCreated') eventData.event_type = 'LogCreated'
+
+    if (process.env.BS_TESTOPS_BUILD_COMPLETED) {
+        if (!process.env.BS_TESTOPS_JWT) {
+            log.debug(`[${logTag}] Missing Authentication Token/ Build ID`)
+            return {
+                status: 'error',
+                message: 'Token/buildID is undefined, build creation might have failed'
+            }
+        }
+        const config = {
+            agent: keepAliveAgent(),
+            headers: {
+                'Authorization': `Bearer ${process.env.BS_TESTOPS_JWT}`,
+                'Content-Type': 'application/json',
+                'X-BSTACK-OBS': 'true'
+            },
+        }
+
+        try {
+            const url = `${DATA_ENDPOINT}/api/v1/event`
+            const data = await got.post(url, { json: eventData, ...config }).json()
+            log.debug(`[${logTag}] Success response: ${require('util').inspect(data, { depth: null })}`)
+        } catch (error) {
+            log.debug(`[${logTag}] Failed. Error: ${error}`)
+        }
+    }
+}
+
+// get hierarchy for a particular test (called by reporter for skipped tests)
+export function getHierarchy(fullTitle?: string) {
+    if (!fullTitle) return []
+    return fullTitle.split('.').slice(0, -1)
+}
+
+export function getFrameworkVersion(framework?: string) {
+    try {
+        if (framework == 'mocha') {
+            return require('../../wdio-mocha-framework/package.json').version
+        } else if (framework == 'cucumber') {
+            return require('../../wdio-cucumber-framework/package.json').version
```

Comment:
This won't work once people download it from NPM given that these dependencies are located differently. I suggest to get the package version somehow else, e.g. using `require.resolve` to get the location of the dependency and then read its `package.json`.

Reply:
Also why is Jasmine missing?